pkg/registry/apis/query/query.go

```diff
@@ -2,92 +2,185 @@ package query
+	// Actually run the query
+	rsp, err := r.execute(ctx, req)
 	if err != nil {
-		errhttp.Write(r.Context(), err, w)
-		return
+		return nil, err
 	}
 
 	statusCode := http.StatusOK
-	for _, res := range qdr.Responses {
+	for _, res := range rsp.Responses {
 		if res.Error != nil {
 			statusCode = http.StatusBadRequest
-			if b.returnMultiStatus {
+			if r.builder.returnMultiStatus {
 				statusCode = http.StatusMultiStatus
 			}
 		}
 	}
 	if statusCode != http.StatusOK {
 		requestmeta.WithDownstreamStatusSource(ctx)
 	}
-
-	w.Header().Set("Content-Type", "application/json")
-	err = json.NewEncoder(w).Encode(qdr)
-	if err != nil {
-		errhttp.Write(r.Context(), err, w)
-	}
+	return &query.QueryDataResponse{
+		QueryDataResponse: *rsp,
+	}, err
 }
 
-// See:
-// https://github.com/grafana/grafana/blob/v10.2.3/pkg/services/query/query.go#L88
-func (b *QueryAPIBuilder) processRequest(ctx context.Context, req parsedQueryRequest) (qdr *backend.QueryDataResponse, err error) {
+func (r *subQueryREST) execute(ctx context.Context, req parsedRequestInfo) (qdr *backend.QueryDataResponse, err error) {
 	switch len(req.Requests) {
 	case 0:
 		break // nothing to do
 	case 1:
-		qdr, err = b.handleQuerySingleDatasource(ctx, req.Requests[0])
+		qdr, err = r.builder.handleQuerySingleDatasource(ctx, req.Requests[0])
 	default:
-		qdr, err = b.executeConcurrentQueries(ctx, req.Requests)
+		qdr, err = r.builder.executeConcurrentQueries(ctx, req.Requests)
 	}
 
 	if len(req.Expressions) > 0 {
-		return b.handleExpressions(ctx, qdr, req.Expressions)
+		qdr, err = r.builder.handleExpressions(ctx, req, qdr)
+	}
+
+	// Remove hidden results
+	for _, refId := range req.HideBeforeReturn {
+		r, ok := qdr.Responses[refId]
+		if ok && r.Error == nil {
+			delete(qdr.Responses, refId)
+		}
 	}
-	return qdr, err
+	return
 }
 
 // Process a single request
 // See: https://github.com/grafana/grafana/blob/v10.2.3/pkg/services/query/query.go#L242
-func (b *QueryAPIBuilder) handleQuerySingleDatasource(ctx context.Context, req groupedQueries) (*backend.QueryDataResponse, error) {
-	gv, err := b.registry.GetDatasourceGroupVersion(req.pluginId)
+func (b *QueryAPIBuilder) handleQuerySingleDatasource(ctx context.Context, req datasourceRequest) (*backend.QueryDataResponse, error) {
+	ctx, span := b.tracer.Start(ctx, "Query.handleQuerySingleDatasource")
+	defer span.End()
+	span.SetAttributes(
+		attribute.String("datasource.type", req.PluginId),
+		attribute.String("datasource.uid", req.UID),
+	)
+
+	allHidden := true
+	for idx := range req.Request.Queries {
+		if !req.Request.Queries[idx].Hide {
+			allHidden = false
+			break
+		}
+	}
+	if allHidden {
+		return &backend.QueryDataResponse{}, nil
+	}
+
+	// headers?
+	client, err := b.client.GetDataSourceClient(ctx, v0alpha1.DataSourceRef{
+		Type: req.PluginId,
+		UID:  req.UID,
+	})
 	if err != nil {
 		return nil, err
 	}
-	return b.runner.ExecuteQueryData(ctx, gv, req.uid, req.query)
+
+	// headers?
```

Comment:
is this an important question to answer before this merges?

Reply:
Not before this merge -- it will be important for how we get auth headers from alerting client code passed along to each of the steps; but I think it is too much to get sorted in this one.